eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Plutus/Context.hs

```diff
@@ -0,0 +1,125 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DefaultSignatures #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableSuperClasses #-}
+
+module Cardano.Ledger.Alonzo.Plutus.Context (
+  EraPlutusTxInfo (..),
+  EraPlutusContext (..),
+  mkPlutusLanguageContext,
+
+  -- * Language dependent translation
+  PlutusTxInfo,
+  PlutusTxCert,
+  PlutusScriptPurpose,
+  PlutusScriptContext,
+)
+where
+
+import Cardano.Ledger.Alonzo.Scripts (AlonzoEraScript, PlutusScript (..))
+import Cardano.Ledger.Alonzo.Tx (ScriptPurpose (..))
+import Cardano.Ledger.Binary (DecCBOR (..), EncCBOR (..))
+import Cardano.Ledger.Core
+import Cardano.Ledger.Plutus.Data (Data (..))
+import Cardano.Ledger.Plutus.Language (Language (..), PlutusLanguage)
+import Cardano.Ledger.UTxO (UTxO (..))
+import Cardano.Slotting.EpochInfo (EpochInfo)
+import Cardano.Slotting.Time (SystemStart)
+import Control.DeepSeq (NFData)
+import Data.Kind (Type)
+import Data.Text (Text)
+import NoThunks.Class (NoThunks)
+import qualified PlutusLedgerApi.V1 as P (ToData, toData)
+import qualified PlutusLedgerApi.V1 as PV1
+import qualified PlutusLedgerApi.V2 as PV2
+import qualified PlutusLedgerApi.V3 as PV3
+
+class (PlutusLanguage l, EraPlutusContext era) => EraPlutusTxInfo (l :: Language) era where
+  toPlutusTxCert :: proxy l -> TxCert era -> Either (ContextError era) (PlutusTxCert l)
+
+  toPlutusScriptPurpose ::
+    proxy l ->
+    ScriptPurpose era ->
+    Either (ContextError era) (PlutusScriptPurpose l)
+
+  toPlutusTxInfo ::
+    proxy l ->
+    PParams era ->
+    EpochInfo (Either Text) ->
+    SystemStart ->
+    UTxO era ->
+    Tx era ->
+    Either (ContextError era) (PlutusTxInfo l)
+
+  toPlutusScriptContext ::
+    proxy l ->
+    PlutusTxInfo l ->
+    ScriptPurpose era ->
+    Either (ContextError era) (PlutusScriptContext l)
+
+class
+  ( AlonzoEraScript era
+  , Eq (ContextError era)
+  , Show (ContextError era)
+  , NFData (ContextError era)
+  , NoThunks (ContextError era)
+  , EncCBOR (ContextError era)
+  , DecCBOR (ContextError era)
+  ) =>
+  EraPlutusContext era
+  where
+  data ContextError era :: Type
+
+  mkPlutusScriptContext ::
```

Comment:
So this function, ` mkPlutusScriptContext `, seems to be the "boss" . 
All the functions from `EraPlutusTxInfo` (  `toPlutusTxCert`, `toPlutusScriptPurpose`, `toPlutusTxInfo`, `toPlutusScriptContext`)   are used to build  `mkPlutusLanguageContext`, which in turn is used in each  `mkPlutusScriptContext ` implementation.  It is the one ultimately used in  in `Evaluate.collectPlutusScriptsWithContext` to build `PlutusWithContext`, which then gets passed to `evalPlutusScripts` in  Utxos rule. 
Not really a question or useful comment, sorry - just validating my understanding.  Anyway, such a clean model compared to `ExtendedUtxo` -txInfo and `valContext - and their unclear (to me) interactions. 
Love how it all fits together so neatly.

Reply:
You are right on point!